### PR TITLE
Feature/9038 delete catalogue item endpoint

### PR DESCRIPTION
**Pull request by kiran-mastek**
Delete a Catalogue Solution

API endpoint

Story: [AB#9038](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/9038)
Task: [AB#12969](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/12969)

### CODE REVIEW

**Review by bjss-andrew — DISMISSED** [review]

File: src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/OrderItemsController.cs [hunk 1]
```diff
@@ -118,5 +118,36 @@ public async Task<ActionResult<GetOrderItemModel>> GetAsync(CallOffId callOffId,
...
+            CallOffId callOffId,
+            CatalogueItemId catalogueItemId)
+        {
+            Expression<Func<Order, IEnumerable<OrderItem>>> orderItems = o =>
+                o.OrderItems.Where(i => i.CatalogueItem.Id == catalogueItemId);
+
+            var order = await context.Order
+                .Where(o => o.Id == callOffId.Id)
+                .Include(orderItems)
+                .ThenInclude(x => x.CatalogueItem)
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/OrderItemsController.cs:134]: We generally avoid using `x` for single letter lambda parameter names. A letter that relates to the parameter is preferable, e.g. `i` might be better here.

File: src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/OrderItemsController.cs [hunk 2]
```diff
@@ -118,5 +118,36 @@ public async Task<ActionResult<GetOrderItemModel>> GetAsync(CallOffId callOffId,
...
+            if (order is null || order.IsDeleted)
+            {
+                return NotFound();
+            }
+
+            var deleted = order.DeleteOrderItem(catalogueItemId);
+            if (!deleted)
+            {
+                return NotFound();
+            }
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/OrderItemsController.cs:146]: I'm not sure we gain anything by introducing the variable here.

File: src/NHSD.BuyingCatalogue.Ordering.Domain/Order.cs [hunk 3]
```diff
@@ -200,6 +200,16 @@ public Result Complete()
             return Result.Success();
         }
 
+        public bool DeleteOrderItem(CatalogueItemId catalogueItemId)
+        {
+            var orderItem = orderItems.SingleOrDefault(x => x.CatalogueItem.Id == catalogueItemId);
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Domain/Order.cs:205]: As above with the lambda parameter name.

File: src/NHSD.BuyingCatalogue.Ordering.Domain/Order.cs [hunk 4]
```diff
@@ -200,6 +200,16 @@ public Result Complete()
...
+            if (orderItem == null)
+                return false;
+
+            orderItems.Remove(orderItem);
+            return true;
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Domain/Order.cs:210]: Can we simplify this to `return orderItems.Remove(orderItem);`?

**Review by bjss-chris — CHANGES_REQUESTED** [review]

File: src/NHSD.BuyingCatalogue.Ordering.Domain/Order.cs [hunk 5]
```diff
@@ -202,12 +202,11 @@ public Result Complete()
 
         public bool DeleteOrderItem(CatalogueItemId catalogueItemId)
         {
-            var orderItem = orderItems.SingleOrDefault(x => x.CatalogueItem.Id == catalogueItemId);
+            var orderItem = orderItems.SingleOrDefault(o => o.CatalogueItem.Id == catalogueItemId);
             if (orderItem == null)
                 return false;
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Domain/Order.cs]: I believe `List<T>.Remove` accepts `null` so we should be able to omit the `if` statement but please correct me if I'm wrong.

**Review by bjss-chris — CHANGES_REQUESTED** [review]

File: src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/OrderItemsController.cs (hunk 1 above)
Comment (kiran-mastek) [reply src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/OrderItemsController.cs]: done

File: src/NHSD.BuyingCatalogue.Ordering.Domain/Order.cs (hunk 3 above)
Comment (kiran-mastek) [reply src/NHSD.BuyingCatalogue.Ordering.Domain/Order.cs]: done

File: src/NHSD.BuyingCatalogue.Ordering.Domain/Order.cs (hunk 4 above)
Comment (kiran-mastek) [reply src/NHSD.BuyingCatalogue.Ordering.Domain/Order.cs]: done

File: src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/OrderItemsController.cs (hunk 2 above)
Comment (kiran-mastek) [reply src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/OrderItemsController.cs]: done

File: src/NHSD.BuyingCatalogue.Ordering.Domain/Order.cs (hunk 5 above)
Comment (kiran-mastek) [reply src/NHSD.BuyingCatalogue.Ordering.Domain/Order.cs]: done with list remove

**Review by bjss-chris — APPROVED** [review]

**Pull request merged**